File: implementations/go/provekit-ir-symbolic/ir/property_test.go
package ir

import (
	"strings"
	"testing"
)

func TestContractCollectsDeclaration(t *testing.T) {
	ResetCollector()
	finish := BeginCollecting()
	Property("zeroIsZero", Eq(ParseInt(StrConst("0")), Num(0)))
	decls := finish()

	if len(decls) != 1 {
		t.Fatalf("want 1 decl, got %d", len(decls))
	}
	cd, ok := decls[0].(ContractDeclaration)
	if !ok {
		t.Fatalf("want ContractDeclaration, got %T", decls[0])
	}
	if cd.Name != "zeroIsZero" {
		t.Errorf("name: want zeroIsZero, got %s", cd.Name)
	}
	if cd.OutBinding != "out" {
		t.Errorf("outBinding: want out, got %s", cd.OutBinding)
	}
	if cd.Pre == nil {
		t.Errorf("Pre should be set by Property()")
	}
}

func TestContractWithPostAndInv(t *testing.T) {
	ResetCollector()
	finish := BeginCollecting()
	pre := Eq(Num(1), Num(1))
	post := Eq(Num(2), Num(2))
	inv := Eq(Num(3), Num(3))
	Contract("threeSlots", ContractArgs{Pre: pre, Post: post, Inv: inv})
	decls := finish()

	cd := decls[0].(ContractDeclaration)
	if cd.Pre == nil || cd.Post == nil || cd.Inv == nil {
		t.Errorf("all three slots should be set")
	}
}

func TestContractRequiresAtLeastOneSlot(t *testing.T) {
	ResetCollector()
	BeginCollecting()
	defer ResetCollector()
	defer func() {
		r := recover()
		if r == nil {
			t.Fatal("want panic when all slots nil, got none")
		}
		msg, _ := r.(string)
		if !strings.Contains(msg, "at least one of Pre / Post / Inv") {
			t.Errorf("panic message: %v", r)
		}
	}()
	Contract("emptySlots", ContractArgs{})
}

func TestMustIsPreconditionAlias(t *testing.T) {
	ResetCollector()
	finish := BeginCollecting()
	Must("aliasOfPre", Eq(Num(0), Num(0)))
	decls := finish()

	cd := decls[0].(ContractDeclaration)
	if cd.Pre == nil {
		t.Errorf("Must should set Pre")
	}
	if cd.Post != nil || cd.Inv != nil {
		t.Errorf("Must should not set Post/Inv")
	}
}

func TestBridgeCollectsDeclaration(t *testing.T) {
	ResetCollector()
	finish := BeginCollecting()
	Bridge("parseIntBridgesV8", BridgeSpec{
		SourceSymbol:      "global.parseInt",
		SourceLayer:       "ts-kit@1.0",
		TargetContractCid: "abc1234567890def",
		TargetLayer:       "V8@12.4",
		Notes:             "the canonical bridge",
	})
	decls := finish()

	if len(decls) != 1 {
		t.Fatalf("want 1 decl, got %d", len(decls))
	}
	bd, ok := decls[0].(BridgeDeclaration)
	if !ok {
		t.Fatalf("want BridgeDeclaration, got %T", decls[0])
	}
	if bd.SourceSymbol != "global.parseInt" {
		t.Errorf("source: want global.parseInt, got %s", bd.SourceSymbol)
	}
	if bd.Notes != "the canonical bridge" {
		t.Errorf("notes: want 'the canonical bridge', got %s", bd.Notes)
	}
}

func TestMultipleDeclsCollectInOrder(t *testing.T) {
	ResetCollector()
	finish := BeginCollecting()
	Property("p1", Eq(Num(0), Num(0)))
	Bridge("b1", BridgeSpec{
		SourceSymbol:      "x",
		SourceLayer:       "L1",
		TargetContractCid: strings.Repeat("0", 32),
		TargetLayer:       "L2",
	})
	Property("p2", Eq(Num(1), Num(1)))
	decls := finish()

	if len(decls) != 3 {
		t.Fatalf("want 3 decls, got %d", len(decls))
	}
	wantKinds := []string{"contract", "bridge", "contract"}
	wantNames := []string{"p1", "b1", "p2"}
	for i, d := range decls {
		if d.Kind() != wantKinds[i] {
			t.Errorf("decls[%d] kind: want %s, got %s", i, wantKinds[i], d.Kind())
		}
		if d.DeclName() != wantNames[i] {
			t.Errorf("decls[%d] name: want %s, got %s", i, wantNames[i], d.DeclName())
		}
	}
}

func TestPropertyOutsideCollectorPanics(t *testing.T) {
	ResetCollector()
	defer func() {
		r := recover()
		if r == nil {
			t.Fatal("want panic, got none")
		}
		if !strings.Contains(r.(string), "outside an active collector") {
			t.Errorf("panic message: %v", r)
		}
	}()
	Property("orphan", Eq(Num(0), Num(0)))
}

func TestNestedBeginCollectingPanics(t *testing.T) {
	ResetCollector()
	finish := BeginCollecting()
	defer finish()
	defer func() {
		r := recover()
		if r == nil {
			t.Fatal("want panic, got none")
		}
		if !strings.Contains(r.(string), "already active") {
			t.Errorf("panic message: %v", r)
		}
	}()
	BeginCollecting()
}

func TestForAllReturnsFlatQuantifier(t *testing.T) {
	ResetCollector()
	BeginCollecting()
	defer ResetCollector()

	f := ForAll(Int, func(x IrTerm) IrFormula {
		return Gt(x, Num(0))
	})
	q, ok := f.(quantFormula)
	if !ok {
		t.Fatalf("want quantFormula, got %T", f)
	}
	if q.Kind != "forall" {
		t.Errorf("kind: want forall, got %s", q.Kind)
	}
	if q.Sort != Int {
		t.Errorf("sort: want Int")
	}
	if _, ok := q.Body.(atomicFormula); !ok {
		t.Errorf("body: want atomicFormula, got %T", q.Body)
	}
}

func TestExistsReturnsFlatQuantifier(t *testing.T) {
	ResetCollector()
	BeginCollecting()
	defer ResetCollector()

	f := Exists(String, func(s IrTerm) IrFormula {
		return Eq(ParseInt(s), Num(0))
	})
	q, ok := f.(quantFormula)
	if !ok {
		t.Fatalf("want quantFormula, got %T", f)
	}
	if q.Kind != "exists" {
		t.Errorf("kind: want exists, got %s", q.Kind)
	}
}

func TestDescribeMustBuildsPath(t *testing.T) {
	ResetCollector()
	finish := BeginCollecting()

	Describe("parseInt", func() {
		Must("canReturnZero", Exists(String, func(s IrTerm) IrFormula {
			return Eq(ParseInt(s), Num(0))
		}))
	})

	decls := finish()
	if len(decls) != 1 {
		t.Fatalf("want 1 decl, got %d", len(decls))
	}
	if decls[0].DeclName() != "parseInt > canReturnZero" {
		t.Errorf("name: want 'parseInt > canReturnZero', got %s", decls[0].DeclName())
	}
}

func TestNestedDescribeBuildsPath(t *testing.T) {
	ResetCollector()
	finish := BeginCollecting()

	Describe("Math", func() {
		Describe("abs", func() {
			Must("non-negative", ForAll(Int, func(x IrTerm) IrFormula {
				return Gt(Abs(x), Num(-1))
			}))
		})
	})

	decls := finish()
	if len(decls) != 1 {
		t.Fatalf("want 1 decl, got %d", len(decls))
	}
	if decls[0].DeclName() != "Math > abs > non-negative" {
		t.Errorf("name: want 'Math > abs > non-negative', got %s", decls[0].DeclName())
	}
}

func TestDescribePopsAfterBody(t *testing.T) {
	ResetCollector()
	finish := BeginCollecting()

	Describe("a", func() {
		Must("inner", Eq(Num(0), Num(0)))
	})
	Must("outer", Eq(Num(1), Num(1)))

	decls := finish()
	got := []string{decls[0].DeclName(), decls[1].DeclName()}
	want := []string{"a > inner", "outer"}
	for i := range got {
		if got[i] != want[i] {
			t.Errorf("decls[%d]: want %s, got %s", i, want[i], got[i])
		}
	}
}

func TestMustSkipIsNoOp(t *testing.T) {
	ResetCollector()
	finish := BeginCollecting()

	Describe("parseInt", func() {
		MustSkip("legacy", Eq(Num(0), Num(0)))
		Must("real", Eq(Num(1), Num(1)))
	})

	decls := finish()
	if len(decls) != 1 {
		t.Fatalf("want 1 decl, got %d", len(decls))
	}
	if decls[0].DeclName() != "parseInt > real" {
		t.Errorf("name: want 'parseInt > real', got %s", decls[0].DeclName())
	}
}

func TestDescribeSkipDoesNotInvokeBody(t *testing.T) {
	ResetCollector()
	finish := BeginCollecting()

	bodyRan := false
	DescribeSkip("never", func() {
		bodyRan = true
		Must("ignored", Eq(Num(0), Num(0)))
	})

	decls := finish()
	if len(decls) != 0 {
		t.Errorf("want 0 decls, got %d", len(decls))
	}
	if bodyRan {
		t.Errorf("body should not have run")
	}
}

func TestResetCollectorRecoversFromLeakedState(t *testing.T) {
	BeginCollecting()
	// Simulate exception leaking the collector.
	ResetCollector()

	finish := BeginCollecting()
	Property("ok", Eq(Num(0), Num(0)))
	decls := finish()
	if len(decls) != 1 {
		t.Fatalf("want 1 decl, got %d", len(decls))
	}
}

func TestQuantifierCounterResetsBetweenCollections(t *testing.T) {
	ResetCollector()

	finish1 := BeginCollecting()
	f1 := ForAll(Int, func(x IrTerm) IrFormula { return Eq(x, Num(0)) })
	_ = f1
	finish1()

	finish2 := BeginCollecting()
	f2 := ForAll(Int, func(x IrTerm) IrFormula { return Eq(x, Num(0)) })
	defer finish2()

	q1 := f1.(quantFormula)
	q2 := f2.(quantFormula)

	if q1.Name != "_x0" {
		t.Errorf("first run name: want _x0, got %s", q1.Name)
	}
	if q2.Name != "_x0" {
		t.Errorf("second run name: want _x0 (counter reset), got %s", q2.Name)
	}
}

File: implementations/rust/provekit-ir-symbolic/tests/property_test.rs
//! Mirror of the collector / describe / must / bridge tests in
//! `src/ir/symbolic/symbolic.test.ts`.

use provekit_ir_symbolic::property::{
    _reset_collector, begin_collecting, BridgeSpec, Declaration,
};
use provekit_ir_symbolic::types::sorts;
use provekit_ir_symbolic::*;
use provekit_ir_symbolic::{describe as describe_macro, exists, forall, must as must_macro, bridge as bridge_macro};

fn reset() { _reset_collector(); }

#[test]
fn property_collects_declaration() {
    reset();
    let handle = begin_collecting();
    property::property("zeroIsZero", eq(parse_int(str_("0")), num(0_i64)));
    let decls = handle.finish();
    assert_eq!(decls.len(), 1);
    match &decls[0] {
        Declaration::Property { name, formula } => {
            assert_eq!(name, "zeroIsZero");
            assert_eq!(serde_json::to_value(formula).unwrap()["kind"], serde_json::json!("atomic"));
        }
        _ => panic!("expected property declaration"),
    }
}

#[test]
fn bridge_collects_declaration() {
    reset();
    let handle = begin_collecting();
    bridge_macro!("parseIntBridgesV8", BridgeSpec {
        source_symbol: "global.parseInt".into(),
        source_layer: "ts-kit@1.0".into(),
        target_contract_cid: "abc1234567890def".into(),
        target_layer: "V8@12.4".into(),
        notes: Some("the canonical bridge".into()),
    });
    let decls = handle.finish();
    assert_eq!(decls.len(), 1);
    match &decls[0] {
        Declaration::Bridge { name, source_symbol, target_contract_cid, notes, .. } => {
            assert_eq!(name, "parseIntBridgesV8");
            assert_eq!(source_symbol, "global.parseInt");
            assert_eq!(target_contract_cid, "abc1234567890def");
            assert_eq!(notes.as_deref(), Some("the canonical bridge"));
        }
        _ => panic!("expected bridge"),
    }
}

#[test]
fn multiple_property_and_bridge_calls_collect_in_order() {
    reset();
    let handle = begin_collecting();
    property::property("p1", eq(num(0_i64), num(0_i64)));
    bridge_macro!("b1", BridgeSpec {
        source_symbol: "x".into(),
        source_layer: "L1".into(),
        target_contract_cid: "0".repeat(32),
        target_layer: "L2".into(),
        notes: None,
    });
    property::property("p2", eq(num(1_i64), num(1_i64)));
    let decls = handle.finish();
    let kinds: Vec<&str> = decls.iter().map(|d| match d {
        Declaration::Property { .. } => "property",
        Declaration::Bridge { .. } => "bridge",
    }).collect();
    assert_eq!(kinds, vec!["property", "bridge", "property"]);
    let names: Vec<&str> = decls.iter().map(|d| d.name()).collect();
    assert_eq!(names, vec!["p1", "b1", "p2"]);
}

#[test]
#[should_panic(expected = "outside an active collector")]
fn calling_property_without_active_collector_panics() {
    reset();
    property::property("x", eq(num(0_i64), num(0_i64)));
}

#[test]
#[should_panic(expected = "already active")]
fn nested_begin_collecting_panics() {
    reset();
    let _h = begin_collecting();
    let _h2 = begin_collecting();
}

#[test]
fn describe_must_registers_full_path_name() {
    reset();
    let handle = begin_collecting();
    describe_macro!("parseInt", {
        must_macro!("canReturnZero",
            exists!(s: sorts::string() => eq(parse_int(s), num(0_i64))));
    });
    let decls = handle.finish();
    assert_eq!(decls.len(), 1);
    assert_eq!(decls[0].name(), "parseInt > canReturnZero");
}

#[test]
fn nested_describes_build_path() {
    reset();
    let handle = begin_collecting();
    describe_macro!("Math", {
        describe_macro!("abs", {
            must_macro!("non-negative",
                forall!(x: sorts::int() => gt(abs(x), num(-1_i64))));
        });
    });
    let decls = handle.finish();
    assert_eq!(decls.len(), 1);
    assert_eq!(decls[0].name(), "Math > abs > non-negative");
}

#[test]
fn describe_pops_segment_after_body() {
    reset();
    let handle = begin_collecting();
    describe_macro!("a", {
        must_macro!("inner", eq(num(0_i64), num(0_i64)));
    });
    must_macro!("outer", eq(num(1_i64), num(1_i64)));
    let decls = handle.finish();
    let names: Vec<&str> = decls.iter().map(|d| d.name()).collect();
    assert_eq!(names, vec!["a > inner", "outer"]);
}

#[test]
fn must_skip_is_noop() {
    reset();
    let handle = begin_collecting();
    describe_macro!("parseInt", {
        property::must_skip("legacy invariant", eq(num(0_i64), num(0_i64)));
        must_macro!("real invariant", eq(num(1_i64), num(1_i64)));
    });
    let decls = handle.finish();
    assert_eq!(decls.len(), 1);
    assert_eq!(decls[0].name(), "parseInt > real invariant");
}

#[test]
fn describe_skip_does_not_run_body() {
    reset();
    let handle = begin_collecting();
    let mut body_ran = 0_i32;
    let body_ran_ref = &mut body_ran;
    property::describe_skip("never", || {
        *body_ran_ref += 1;
    });
    let decls = handle.finish();
    assert_eq!(decls.len(), 0);
    assert_eq!(body_ran, 0);
}

#[test]
fn forall_wraps_body_builder() {
    reset();
    let f = forall!(x: sorts::int() => gt(x, num(0_i64)));
    let v = serde_json::to_value(&f).unwrap();
    assert_eq!(v["kind"], serde_json::json!("forall"));
    assert_eq!(v["sort"]["name"], serde_json::json!("Int"));
    assert_eq!(v["predicate"]["body"]["kind"], serde_json::json!("atomic"));
}

#[test]
fn exists_wraps_body_builder() {
    reset();
    let f = exists!(s: sorts::string() => eq(parse_int(s), num(0_i64)));
    assert_eq!(serde_json::to_value(&f).unwrap()["kind"], serde_json::json!("exists"));
}

#[test]
fn nested_quantifiers_compose() {
    reset();
    let f = forall!(x: sorts::int() =>
        forall!(y: sorts::int() => gt(add(x.clone(), y), num(0_i64))));
    assert_eq!(serde_json::to_value(&f).unwrap()["kind"], serde_json::json!("forall"));
}

#[test]
fn reset_collector_lets_new_begin_collecting_succeed_after_leak() {
    reset();
    let _leaked = begin_collecting();
    std::mem::forget(_leaked); // simulate the panic-leak path: handle dropped without finishing
    reset();
    let handle = begin_collecting();
    property::property("ok", eq(num(0_i64), num(0_i64)));
    let decls = handle.finish();
    assert_eq!(decls.len(), 1);
}
